runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/GroupingShuffleReader.java

```diff
@@ -402,6 +432,14 @@ public boolean hasNext() {
 
       @Override
       public V next() {
+        // Given that the underlying ReadOperation already checks the abort status after every
+        // record it advances over (i.e., for every distinct key), we skip the check when at
+        // the first value as that is redundant. Signal by thread interruption may be better, but
+        // it may also have unintended side-effects.
+        if (!atFirstValue && aborted.get()) {
```

Comment:
Checking atFirstValue and aborted will likely perform worse then just checking aborted all the time. It may seem redundant but the abort happens asynchronously so we may have gotten past the check in the ReadOperation already.